fish-rust/src/ast.rs

```diff
@@ -0,0 +1,5206 @@
+        &mut self,
+        _node: &mut Option<DecoratedStatementDecorator>,
+    ) {
+    }
+    fn visit_job_conjunction_decorator(&mut self, _node: &mut Option<JobConjunctionDecorator>) {}
+    fn visit_else_clause(&mut self, _node: &mut Option<ElseClause>) {}
+    fn visit_semi_nl(&mut self, _node: &mut Option<SemiNl>) {}
+    fn visit_time(&mut self, _node: &mut Option<KeywordTime>) {}
+    fn visit_token_background(&mut self, _node: &mut Option<TokenBackground>) {}
+}
+
+pub trait AcceptorMut {
+    fn accept_mut(&mut self, visitor: &mut dyn NodeVisitorMut, reversed: bool);
+}
+
+impl<T: AcceptorMut> AcceptorMut for Option<T> {
+    fn accept_mut(&mut self, visitor: &mut dyn NodeVisitorMut, reversed: bool) {
+        match self {
+            Some(node) => node.accept_mut(visitor, reversed),
+            None => (),
+        }
+    }
+}
+
+/// Node is the base trait of all AST nodes.
+pub trait Node: Acceptor + AcceptorMut + std::fmt::Debug {
+    /// The parent node, or null if this is root.
+    fn parent(&self) -> Option<&dyn Node>;
+    fn parent_ffi(&self) -> &Option<*const dyn Node>;
+
+    /// The type of this node.
+    fn typ(&self) -> Type;
+
+    /// The category of this node.
+    fn category(&self) -> Category;
+
+    /// \return a helpful string description of this node.
+    #[widestrs]
+    fn describe(&self) -> WString {
+        let mut res = ast_type_to_string(self.typ()).to_owned();
+        if let Some(n) = self.as_token() {
+            let token_type: &'static wstr = n.token_type().into();
+            res += &sprintf!(" '%ls'"L, token_type)[..];
+        } else if let Some(n) = self.as_keyword() {
+            let keyword: &'static wstr = n.keyword().into();
+            res += &sprintf!(" '%ls'"L, keyword)[..];
+        }
+        res
+    }
+
+    /// \return the source range for this node, or none if unsourced.
+    /// This may return none if the parse was incomplete or had an error.
+    fn try_source_range(&self) -> Option<SourceRange>;
+
+    /// \return the source range for this node, or an empty range {0, 0} if unsourced.
+    fn source_range(&self) -> SourceRange {
+        self.try_source_range().unwrap_or(SourceRange::new(0, 0))
+    }
+
+    /// \return the source code for this node, or none if unsourced.
+    fn try_source<'s>(&self, orig: &'s wstr) -> Option<&'s wstr> {
+        self.try_source_range()
+            .map(|r| &orig[r.start as usize..r.end() as usize])
+    }
+
+    /// \return the source code for this node, or an empty string if unsourced.
+    fn source<'s>(&self, orig: &'s wstr) -> &'s wstr {
+        self.try_source(orig).unwrap_or_default()
+    }
+
+    // The address of the object, for comparison.
+    fn as_ptr(&self) -> *const ();
+
+    // Cast to any sub-trait.
+
+    fn as_keyword(&self) -> Option<&dyn Keyword> {
+        None
+    }
+    fn as_leaf(&self) -> Option<&dyn Leaf> {
+        None
+    }
+    fn as_token(&self) -> Option<&dyn Token> {
+        None
+    }
```

Comment:
Depending on how these are used, you might have another option. You can (and this is the idiomatic rust way where possible) avoid returning an `Option<T>` at all by implementing these trait methods (here) only for types that implement the sub-trait by just putting a generic constraint on the default method impl (but then you can't just blindly call them on any random node to introspect whether or not it implements the trait if you're doing it all via dynamic dispatch and never via a concrete type).